rfc/001-block-retention.md

```diff
@@ -0,0 +1,111 @@
+
+* The majority of nodes should retain blocks longer than the Cosmos SDK unbonding period, for light client verification.
+
+* Some nodes must take and serve state sync snapshots with snapshot intervals less than the block retention periods, to allow new nodes to state sync and then replay blocks to catch up.
+
+* Applications may not persist their state on commit, and require block replay on restart.
+
+* Only a minority of nodes can be state synced within the unbonding period, for light client verification and to serve block histories for catch-up.
+
+However, it is unclear if and how we should enforce this. It may not be possible to technically enforce all of these without knowing the state of the entire network, but it may also be unrealistic to expect this to be enforced entirely through social coordination. This is especially unfortunate since the consequences of misconfiguration can be permanent chain-wide data loss.
+
+## Proposal
+
+Add a new field `retain_height` to the ABCI `ResponseCommit` message:
+
+```proto
+service ABCIApplication {
+  rpc Commit(RequestCommit) returns (ResponseCommit);
+}
+
+message RequestCommit {}
+
+message ResponseCommit {
+  // reserve 1
+  bytes  data          = 2; // the Merkle root hash
+  uint64 retain_height = 3; // the oldest block height to retain
+}
+```
+
+Upon ABCI `Commit`, which finalizes execution of a block in the state machine, Tendermint removes all data for heights lower than `retain_height`. This allows the state machine to control block retention, which is preferable since only it can determine the significance of historical blocks. By default (i.e. with `retain_height=0`) all historical blocks are retained.
+
+Removed data includes not only blocks, but also headers, commit info, consensus params, validator sets, and so on. In the first iteration this will be done synchronously, since the number of heights to remove for each run is assumed to be small (often 1) in the typical case. It can be made asynchronous if this is shown to be necessary, but if we then add support for the state machine to query Tendermint (which is planned) then this will be a source of non-determinism.
+
+Since `retain_height` is dynamic, it is possible for it to refer to a height which has already been removed. For example, commit at height 100 may return `retain_height=90` while commit at height 101 may return `retain_height=80`. This is allowed, and will be ignored - it is the application's responsibility to return appropriate values.
+
+## Open Questions
+
+* Should we add a local config option for Tendermint to retain blocks beyond ABCI `retain_height`? This is needed e.g. for archive nodes. For example, an option `retain_blocks=100` would cause Tendermint to retain at minimum the 100 most recent blocks, but more if ABCI commit `retain_height` requires it. Alternatively, this could be configured in the application.
```

Comment:
I'd prefer one way of doing things

Reply:
I think that makes sense. This would also give the app total knowledge about block retention for correct nodes, which might be useful.